bluetooth: host: Discard prohibit field in adv report

LingaoM opened the pull request:
if data_status field is reserved, should discard this adv report.

LingaoM committed:
bluetooth: host: Discard prohibit field in adv report

if data_status field is reserved, should discard this adv report.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
The reason is that we used HCI UART and H4 protocol, and found that there was some electrical interference on the serial that caused abnormal data, leading to problems with the Host. Overall, adding this check will prevent crashes at this location.
``` C
		/* No more data coming from the controller.
		 * Create event.
		 */
		__ASSERT_NO_MSG(is_report_complete);
```

jhedberg commented:
Isn't the code missing a check for `evt_type ==  BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_RX_FAILED`? Is it perhaps that value that you get (which would also cause `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type)` to give you `0x03`).

That said, what you're describing sounds like a workaround. The H:4 transport is supposed to be reliable (e.g. with HW flow control). If it's not reliable then you're using the wrong transport, and you should consider using H:5 instead.

LingaoM committed:
bluetooth: host: Make reassembling state not in reassembling

After https://github.com/zephyrproject-rtos/zephyr/pull/74639

If we receive corrupted adv report, will set state to discarding,
but if next adv report is new advertiser, will lead ASSERT.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
> Isn't the code missing a check for `evt_type == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_RX_FAILED`? Is it perhaps that value that you get (which would also cause `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type)` to give you `0x03`).

But if evt_type is 0bx11xxxxx, not only 0xff, will lead data_status to 0b11

> That said, what you're describing sounds like a workaround. The H:4 transport is supposed to be reliable (e.g. with HW flow control). If it's not reliable then you're using the wrong transport, and you should consider using H:5 instead.

I'm not sure if it's appropriate to mention H4 here, but I want to say that adding this check here doesn't break the protocol.

hermabe commented:
> Isn't the code missing a check for `evt_type == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_RX_FAILED`? Is it perhaps that value that you get (which would also cause `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type)` to give you `0x03`).
> 

This data status is only used by PAwR when failing to receive an AUX_SYNC_SUBEVENT_IND/RSP PDU.



> That said, what you're describing sounds like a workaround. The H:4 transport is supposed to be reliable (e.g. with HW flow control). If it's not reliable then you're using the wrong transport, and you should consider using H:5 instead.

Agree, we expect hci to be reliable, if it is not I expect everything to fall apart.

jhedberg commented:
> I'm not sure if it's appropriate to mention H4 here, but I want to say that adding this check here doesn't break the protocol.

The new check itself seems alright, however if the underlying issue why you need it is an unreliable HCI transport, then you're just papering over it, since corruption could happen for any HCI packet.

LingaoM commented:
> The new check itself seems alright, however if the underlying issue why you need it is an unreliable HCI transport, then you're just papering over it, since corruption could happen for any HCI packet.

It's true, considering that these modifications have a certain degree of generality, this PR is proposed. Overall, we also hope that the Host can be more robust, as discarding a advertising report has a smaller impact than crashing.

jhedberg commented on subsys/bluetooth/host/scan.c:
```diff
@@ -824,6 +824,12 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
 			return;
 		}
 
+		if (data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_RESERVED) {
+			LOG_ERR("Prohibit data status field in advertising report.");
+			reset_reassembling_advertiser();
```
Instead of doing this, shouldn't you go to discarding state?

jhedberg commented on subsys/bluetooth/host/scan.c:
```diff
@@ -874,7 +880,7 @@ void bt_hci_le_adv_ext_report(struct net_buf *buf)
 			 * this is the first report from the new advertiser.
 			 * Initialize the new advertiser.
 			 */
-			__ASSERT_NO_MSG(reassembling_advertiser.state == FRAG_ADV_INACTIVE);
+			__ASSERT_NO_MSG(reassembling_advertiser.state != FRAG_ADV_REASSEMBLING);
```
This seems like a side effect of not going to discarding state.

LingaoM replied:
Don't think so, if set `FRAG_ADV_DISCARDING`and return, and then receive new advertising, will enter here:
```C
		if (is_new_advertiser) {
			/* We are not reassembling reports from an advertiser and
			 * this is the first report from the new advertiser.
			 * Initialize the new advertiser.
			 */
			__ASSERT_NO_MSG(reassembling_advertiser.state == FRAG_ADV_INACTIVE);
			init_reassembling_advertiser(&evt->addr, evt->sid);
		}

```

LingaoM replied:
@jhedberg Ping.
